third_party/move/move-compiler/src/inlining/translate.rs

```diff
@@ -173,7 +177,7 @@ impl<'l, 'r> Visitor for OuterVisitor<'l, 'r> {
             UnannotatedExp_::ModuleCall(mcall) => {
                 if let Some(repl) = self.inliner.module_call(ex.exp.loc, mcall.as_mut()) {
                     ex.exp.value = repl;
-                    VisitorContinuation::Stop
+                    VisitorContinuation::Descend
```

Comment:
Something is not right with this solution. if we Descend, we traverse the tree potentially many times. As a matter of fact, the `SubstitutionVisitor` should do this job for us. (See the comment on line 210: "[The SubstitutionVisitor] recursively invokes itself if it encounters further inlined functions or lambdas.")

Basically what should happen is that the body is constructed as something along this line:

```
let (p1, ..., pn) = (arg1, ..., argn);
body_of_inlined_function
```

... and then the `SubstitutionVisitor` is called on this guy. For some reason, the 'arg' expressions above are not expanded. But there is evidence that macro expansion works for `body_of_inlined_function`, since we have many examples where one inline function calls another (e.g. `vector.move`) and in those cases the called inline functions are expanded just fine.

What could be the reason that `let (p1, ..., pn) = (arg1, ..., argn)` is not correctly expanded? Perhaps this construct is not correctly handled by the visitor logic?

Reply:
The place where the SubstitutionVisitor continues expansion is at line 235:

```
          UnannotatedExp_::ModuleCall(mcall) => {
                if let Some(repl) = self.inliner.module_call(ex.exp.loc, mcall) {
                    ex.exp.value = repl;
                }
                VisitorContinuation::Descend
            },
```

Reply:
I agree, this solution might do some redundant work, although it seemed modest since it doesn't do much unless there is a call to an inlined function in the tree, and the tree is bounded in size.

I'll have to understand the details of the code better more to be more precise.  Hold on.

Reply:
Since you have 3 uses of the Visitor, it's hard to change the definition, but the problem is here in visitor.rs's `Dispatcher::exp`:
```
    pub fn exp(&mut self, ex: &mut Exp) {
        self.type_(&mut ex.ty);
        if self.visitor.exp(ex) == VisitorContinuation::Stop {
            return;
        }
        self.exp_unannotated(ex.exp.loc, &mut ex.exp.value)
    }
```
so in the inlining pass (OuterVisitor) the visitor is called on the method call *before* the subexpressions have been processed.  Returning `VisitorContinuation::Stop` prevents the subexpressions being processed.  What we want for inlining is a "post-order" traversal of the AST, so that the ModuleCall is processed before it turns into something else through inlining.

Interestingly, lambdas are handled differently: they get inlined in the later substitution pass.  After inlining a lambda body into a VarCall, they reprocess the resulting expression (`SubstitutionVisitor::exp` calls `SubstitutionVisitor::var_call`, then returns VisitorContinuation::Descend), which is good as it's not clear whether substitution has happened on the lambda before then.

Reply:
But the SubstitutionVisitor which handles the lambdas does also handle further inline function calls. This is what I was pointing to with line 235: `self.inliner.module_call` handles the expansion. Otherwise inlined functions in the body of an expansion wouldn't work (and they are positively working and tested).

There is something going wrong with the arguments. Again, conceptually the following expression is constructed for  inlining and then SubstitionVisitor called on it:

```
let (p1, ..., pn) = (arg1, ..., argn); // arguments to inlined functions
body_of_inlined_function
```

Why does the current code handle inlining fine in `body_of_inlined_function` but not in `arg1,...`?

Reply:
As discussed in Slack:

A nested inline call f1(f2(3)) in any **non-inlined function** (with both f1 and f2 inlined) fails, while putting the expression in the body of an inlined function is ok.

Incidentally, the error generated differs depending on whether the inlined function is in another module or not.  If in the same module, there is an explicit error (e.g., "IR ERROR: unbound function Self.f2") whereas the cross-module case yields the compiler crash with an unwrap() error.

I've now fixed both cases, and included the relevant tests.